kategory/src/main/kotlin/kategory/arrow/PartialFunction.kt

```diff
@@ -0,0 +1,53 @@
+package kategory
+
+abstract class PartialFunction<in A, out B> : (A) -> B {
+    abstract fun isDefinedAt(a: A): Boolean
+}
+
+fun <A, B> PartialFunction<A, B>.orElse(f: PartialFunction<A, B>): PartialFunction<A, B> {
+    val fthis = this
+    return object : PartialFunction<A, B>() {
+
+        override fun isDefinedAt(a: A): Boolean =
+                fthis.isDefinedAt(a) || f.isDefinedAt(a)
+
+        override fun invoke(x: A): B =
+                if (fthis.isDefinedAt(x)) {
+                    fthis(x)
+                } else {
+                    f(x)
+                }
+
+    }
+}
+
+fun <A, B, C> PartialFunction<A, B>.andThen(f: (B) -> C): PartialFunction<A, C> {
+    val fthis = this
+    return object : PartialFunction<A, C>() {
+        override fun isDefinedAt(a: A): Boolean = fthis.isDefinedAt(a)
+        override fun invoke(a: A): C = f(fthis(a))
+    }
+}
+
+fun <A : Any?, B> case(ff: Tuple2<(A) -> Boolean, (A) -> B>): PartialFunction<A, B> =
+        object : PartialFunction<A, B>() {
+            override fun isDefinedAt(a: A): Boolean = ff.a(a)
+            override fun invoke(a: A): B = ff.b(a)
+        }
+
+inline fun <reified A> typeOf() : (A) -> Boolean = {
+    Try({ it as A }).fold({ false }, { true })
+}
+
+infix fun <A, B> ((A) -> Boolean).then(f: (A) -> B): Tuple2<(A) -> Boolean, (A) -> B> = Tuple2(this, f)
```

Comment:
The issue is that for those to be scoped they would need to be placed inside the companion object. and would have to be individually imported:
```
import kategory.match.dsl.case
import kategory.match.dsl.typeOf
...
```
I have never been able to import members of companions with wildcards in Kotlin and it feels to me that those extra imports are an unnecessary burden to the user. Also this will be the first time we require additional imports beside `kategory.*` are we sure we want to do that?

Alternatively I can create a new `kategory.cases` package for these functions but also the users would have to explicitly import `import kategory.cases.*`

Thoughts?